### ISSUE TITLE
After running for a while aws_starter_demo looses connectivity with AWS IoT

### ISSUE THREAD

**Issue by dimkir** [marvell-iot/aws_starter_sdk#28]
When running the original `aws_starter_demo` included with the device,  it works properly for the first 15 minutes or so (after I have successfully configured AWS IoT credentials and home WiFi).

It works fine as you can see from the console output - I am able to publish the state of the buttons and also receive shadow update requests which  switch the led on and off. 

### Output of serial console

```
Build Time: Nov  4 2015 14:24:25

#### AWS STARTER DEMO ####

Connected successfully to the configured network
Cloud Started
delta topic $aws/things/marvell1/shadow/update/delta
Publishing '{"state": {"reported":{"pb":0,"pb_lambda":0,"led":0}}}' to AWS
Publishing '{"state": {"reported":{"led":1}}}' to AWS
Publishing '{"state": {"reported":{"led":0}}}' to AWS
Publishing '{"state": {"reported":{"led":1}}}' to AWS
Publishing '{"state": {"reported":{"led":0}}}' to AWS
Publishing '{"state": {"reported":{"led":1}}}' to AWS
Publishing '{"state": {"reported":{"led":0}}}' to AWS
Publishing '{"state": {"reported":{"led":1}}}' to AWS
Publishing '{"state": {"reported":{"pb_lambda":1}}}' to AWS
Sending property failed
Publishing '{"state": {"reported":{"pb":1}}}' to AWS
Sending property failed
Publishing '{"state": {"reported":{"pb":2}}}' to AWS
Sending property failed
Publishing '{"state": {"reported":{"pb_lambda":2}}}' to AWS
Sending property failed
Publishing '{"state": {"reported":{"pb":3}}}' to AWS
Sending property failed
Publishing '{"state": {"reported":{"pb":4}}}' to AWS
Sending property failed
Publishing '{"state": {"reported":{"pb_lambda":3}}}' to AWS
Sending property failed
Publishing '{"state": {"reported":{"pb":5}}}' to AWS
Sending property failed
Publishing '{"state": {"reported":{"pb_lambda":4}}}' to AWS
Sending property failed

```

However after like 15 minutes, it seems like a connection drops and I cannot anymore send state of buttons to AWS.

You can see that the `Sending property failed` message popups every time I press the button and attempt to send new state to AWS.
 
Also AWS IoT fails to synchronize shadows with the device - when I update shadow - it comes up as "out of sync" in AWS console. 

Obviously there's loss of connectivity, but is it on WiFi level or on MQTTClient level? 
Eg. in example of WiFi connection loss one of these two callback functions should have been called (see below), but I can see no error messages from these callbacks...

```
void wlan_event_normal_link_lost(void *data)
{
	/* led indication to indicate link loss */
        wmprintf("wlan_event_normal_link_lost()\r\n");
	aws_iot_shadow_disconnect(&mqtt_client);
	device_state = AWS_DISCONNECTED;
}

void wlan_event_normal_connect_failed(void *data)
{
	/* led indication to indicate connect failed */
        wmprintf("wlan_event_normal_connect_failed()\r\n");
	aws_iot_shadow_disconnect(&mqtt_client);
	device_state = AWS_DISCONNECTED;
}
```

**Comment by nikitanerkar** [comment]
Even we have observed the behaviour you have mentioned. As you have correctly stated, the application is not receiving link loss event as the functions are not getting invoked, and therefore the issue is not with WiFi connectivity. We are further investigating the issue. Meanwhile, to fix this behaviour, on receiving 'Sending property failed' error, you can close the existing connection and request for new connection with the AWS IoT.

**Comment by spasse** [comment]
My solution isn't pretty, but it changes my uptime from several minutes to days.  The problem
appears to be that the timer passed to readPacket() sometimes is set to 0.  I just blindly
change it to 10.  Note that there appear to be other instances where it is supposed to be
less than 10 (1?), but always setting it to 10 seems to do no harm...

sdk/external/aws_iot/aws_mqtt_embedded_client_lib/MQTTClient-C/src/MQTTClient.c

int readPacket(Client* c, Timer* timer) 
{
    int rc = MQTT_FAILURE;
    MQTTHeader header = {0};
    int len = 0;
    int rem_len = 0;
    /* 1. read the header byte.  This has the packet type in it */
#if PATCH_BUG
    int left = left_ms(timer);
    if (left != 10) {
	if (left == 0)
	    wmprintf(">>> left: %d\r\n", left);
        left = 10;
    }
    if (c->ipstack->mqttread(c->ipstack, c->readbuf, 1, left) != 1) {
        goto exit;
    }
#else
    if (c->ipstack->mqttread(c->ipstack, c->readbuf, 1, _ms(timer)) != 1)
        goto exit;
#endif // PATCH_BUG

**Comment by anujdeshpande** [comment]
Hi @spasse Looks like you are using an older version of the SDK as well as the associated AWS agent. We have since updated the SDK and more importantly the AWS agent (it's now at 2.1)
https://github.com/marvell-iot/aws_starter_sdk/tree/master/sdk/external/aws_iot

cc @nikitanerkar

**Comment by spasse** [comment]
Hi,

This is a commercial product requiring the use of wmsdk-3.3.33.  I will
upgrade to the external/aws_iot and compare results.

Steve

**Comment by anujdeshpande** [comment]
@spasse If you face any issues doing that, please send a request to anuj AT marvell DOT com and I'll provide you a patch for doing that.

**Comment by spasse** [comment]
Anuj,

I upgraded the sdk to wmsdk-3.4.6
I replaced sdk/external/aws_iot with the 2.1 version from the gitHub site
as of Sept. 2
Modified my program to deal with function signature changes, etc.
Ran overnite on 3 boards.

2 of the 3 boards were in the:

Publishing '{"state": {"reported":{"pb":5}}}' to AWS
Sending property failed

state!

1 of the 3 boards was still connected and working properly.

This is WITHOUT my patches.

Steve
